cpp/src/neighbors/batch_ann/batch_ann.cuh

```diff
@@ -0,0 +1,476 @@
+                                                              params.metric,
+                                                              ivf_pq_params);
+  } else {
+    RAFT_FAIL("Batch KNN build algos only supporting NN Descent and IVF PQ");
+  }
+}
+
+template <typename T, typename IdxT>
+void single_gpu_batch_build(const raft::resources& handle,
+                            raft::host_matrix_view<const T, IdxT, row_major> dataset,
+                            detail::batch_ann_builder<T, IdxT>& knn_builder,
+                            raft::managed_matrix_view<IdxT, IdxT> global_neighbors,
+                            raft::managed_matrix_view<T, IdxT> global_distances,
+                            size_t max_cluster_size,
+                            size_t n_clusters,
+                            batch_ann::index<IdxT, T>& index,
+                            const index_params& batch_params,
+                            raft::host_vector_view<IdxT, IdxT, row_major> cluster_sizes,
+                            raft::host_vector_view<IdxT, IdxT, row_major> cluster_offsets,
+                            raft::host_vector_view<IdxT, IdxT, row_major> inverted_indices)
+{
+  size_t num_rows = dataset.extent(0);
+  size_t num_cols = dataset.extent(1);
+
+  auto cluster_data = raft::make_host_matrix<T, IdxT, row_major>(max_cluster_size, num_cols);
+
+  knn_builder.prepare_build(dataset);
+
+  for (size_t cluster_id = 0; cluster_id < n_clusters; cluster_id++) {
+    size_t num_data_in_cluster = cluster_sizes(cluster_id);
+    size_t offset              = cluster_offsets(cluster_id);
+    if (num_data_in_cluster < static_cast<size_t>(index.k())) {
+      // for the unlikely event where clustering was done lopsidedly and this cluster has less than
+      // k data
+      continue;
+    }
+
+    // dathering dataset from host
+    // cluster_data is on host for now because NN Descent allocates a new device matrix for data of
+    // half fp regardless of whether data is on host or device
+#pragma omp parallel for
+    for (size_t i = 0; i < num_data_in_cluster; i++) {
+      for (size_t j = 0; j < num_cols; j++) {
+        size_t global_row  = inverted_indices(offset + i);
+        cluster_data(i, j) = dataset(global_row, j);
+      }
+    }
+
+    auto cluster_data_view = raft::make_host_matrix_view<const T, IdxT>(
+      cluster_data.data_handle(), num_data_in_cluster, num_cols);
+    auto inverted_indices_view = raft::make_host_vector_view<IdxT, IdxT>(
+      inverted_indices.data_handle() + offset, num_data_in_cluster);
+
+    knn_builder.build_knn(handle,
+                          batch_params,
+                          cluster_data_view,
+                          inverted_indices_view,
+                          global_neighbors,
+                          global_distances);
+  }
+}
+
+template <typename T, typename IdxT>
+void multi_gpu_batch_build(const raft::resources& handle,
```

Comment:
Very nice- I like that you're dispatching instead of providing two different public APIs.